### PR TITLE
Multi-region custom domain name (for easier failover)

### PR DESCRIPTION
**Pull request by forevermatt**
### Added
- Add `serverless-api-dns-for-failover` local module
  * Uses `fail-over-cname` and `custom-domains` to set up custom domain name on both region's APIs, plus simple CNAME record that can be pointed to either one.
- Add `fail-over-cname` local module
- Add `custom-domains` local module
  * Uses `certificate-and-domain` to set up a custom domain and matching https certificate in both regions
- Add `certificate-and-domain` local module
  * Uses new [`terraform-aws-acm-certificate`](https://github.com/silinternational/terraform-aws-acm-certificate) and [`terraform-aws-api-gateway-custom-domain`](https://github.com/silinternational/terraform-aws-api-gateway-custom-domain) modules

### CODE REVIEW

**Comment by forevermatt** [comment]
Once this has been proved and any issues worked out, I'd like to extract the local modules to a separate `terraform-aws-serverless-api-dns-for-failover` module, for us with some of our other Serverless APIs.

File: terraform/main.tf
Comment (Baggerone) [inline terraform/main.tf:84]: Maybe add the version?

File: terraform/modules/certificate-and-domain/versions.tf
Comment (Baggerone) [inline terraform/modules/certificate-and-domain/versions.tf:11]: In one of the idp-terraform files, we use `"~> 3.0"`
Does that matter? Does the difference between `~>` and `>=` matter for either of these?

**Review by Baggerone — COMMENTED** [review]
I didn't get through much of this.  (Steve might have time to approve it.)

File: terraform/main.tf
Comment (briskt) [reply terraform/main.tf:84]: Local modules aren't versioned, so that wouldn't be an option here.

File: terraform/modules/serverless-api-dns-for-failover/main.tf
Comment (briskt) [inline terraform/modules/serverless-api-dns-for-failover/main.tf:1]: You can use `locals.aws_region` and `locals.aws_region.secondary` instead of declaring variables. Eliminates the possibility of the provider and variables getting out of sync. (The same can be done in the serverless-api-dns-for-failover module.)

https://registry.terraform.io/providers/hashicorp/aws/4.8.0/docs/data-sources/region

```suggestion
locals {
  aws_region           = data.aws_region.primary.name
  aws_region_secondary = data.aws_region.secondary.name
}

data "aws_region" "primary" {}
data "aws_region" "secondary" {
  provider = aws.secondary
}

```

File: terraform/modules/custom-domains/main.tf
Comment (briskt) [inline terraform/modules/custom-domains/main.tf:9]: This line is optional since you're passing the default to the default.

File: terraform/modules/custom-domains/versions.tf
Comment (briskt) [inline terraform/modules/custom-domains/versions.tf:9]: The example [here](https://developer.hashicorp.com/terraform/language/providers/configuration#alias-multiple-provider-configurations) says "The following example declares both the `mycloud` and `mycloud.alternate` provider configuration names" and doesn't include the default provider name. I don't know if it will ignore this or do something unexpected.
```suggestion
```

File: terraform/modules/fail-over-cname/outputs.tf
Comment (briskt) [inline terraform/modules/fail-over-cname/outputs.tf:1]: I'd be fine with omitting this file. Your call, though.

**Review by briskt — CHANGES_REQUESTED** [review]

File: terraform/modules/certificate-and-domain/versions.tf
Comment (forevermatt) [reply terraform/modules/certificate-and-domain/versions.tf:11]: Those version constraints are meaningfully different... good observation. I went with the `>=` based on the Terraform documentation about versioning providers within modules, which (surprisingly) says...

> If you are writing a shared Terraform module, constrain only the minimum required provider version using a >= constraint. This should specify the minimum version containing the features your module relies on, and thus allow a user of your module to potentially select a newer provider version if other features are needed by other parts of their overall configuration.

([source](https://developer.hashicorp.com/terraform/language/modules/develop/providers#provider-aliases-within-modules))

File: terraform/modules/serverless-api-dns-for-failover/main.tf
Comment (forevermatt) [reply terraform/modules/serverless-api-dns-for-failover/main.tf:1]: Great suggestion. Thanks!

File: terraform/modules/custom-domains/versions.tf
Comment (forevermatt) [reply terraform/modules/custom-domains/versions.tf:9]: Ahh, good to know. Thanks. 👍

**Review by briskt — APPROVED** [review]
I'll go ahead and approve this and you can merge it when you're ready.

File: terraform/modules/custom-domains/main.tf
Comment (forevermatt) [reply terraform/modules/custom-domains/main.tf:9]: I think I'll leave it (unless you feel strongly otherwise) to help keep the differences between these two module usages all explicit (rather than having to know the defaults). I feel like it helps make it obvious what the differences are between the two `module` blocks in this file.

**Pull request merged**
